### PR TITLE
Migrate ManagedAgent state change to task

### PR DESCRIPTION
**Pull request by fierlion**
### Summary
This migrates the ManagedAgent state change events from the Container state change events to the Task-level state change events to support bugfix model changes.

### Implementation details
This adds batched ManagedAgent events to the Task event

### Testing
I tested manually by building agent/init on an exec-capable container-instance and running 2 separate tasks, one with a single container and another with 3 containers
```
aws ecs start-task \
    --endpoint-url <gamma endpoint> 
    --cluster gammaDefault \
    --task-definition exec-task:2 \
    --container-instances <instance-id> \
    --enable-execute-command
```
followed by:
```
aws ecs describe-tasks \
    --cluster gammaDefault \
    --endpoint-url <gamma endpoint> \
    --tasks <task-id> \
    --debug
```
I got the following output:
```
                    "managedAgents": [
                        {
                            "lastStartedAt": 1607631456.641,
                            "lastStatus": "RUNNING",
                            "name": "ExecuteCommandAgent",
                            "reason": "Execute Command Agent started"
                        }
                    ],
```

New tests cover the changes: Yes.

### Description for the changelog
Migrate ManagedAgent state change from container to task event.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/ecsclient/client.go
Comment (sharanyad) [inline agent/api/ecsclient/client.go:440]: can we do this check at the start, so we dont start building the payload?

File: agent/api/ecsclient/client.go
Comment (sharanyad) [inline agent/api/ecsclient/client.go:446]: debug statement?

File: agent/api/statechange.go
Comment (sharanyad) [inline agent/api/statechange.go:34]: lets use https://github.com/aws/amazon-ecs-agent/blob/cf1a25311ff546b4ab77af57d90cf06c8c8498d4/agent/engine/execcmd/manager.go#L33 instead

File: agent/api/statechange.go
Comment (sharanyad) [inline agent/api/statechange.go:189]: `%v` . or may be just the container name is good enough

File: agent/ecs_client/model/api/api-2.json
Comment (sharanyad) [inline agent/ecs_client/model/api/api-2.json:2118]: is `ManagedAgentStateChanges` already in here?

File: agent/api/ecsclient/client.go
Comment (angelcar) [inline agent/api/ecsclient/client.go:446]: probably a debug level statement?

File: agent/api/statechange.go
Comment (angelcar) [inline agent/api/statechange.go:161]: I created this function because it was reused by container state change and managed state change, but I believe this is invoked in a single place now? If so, it might be better to just move it to container state change creation funciton.

File: agent/api/statechange.go
Comment (angelcar) [inline agent/api/statechange.go:187]: hmm, maybe it is better to not assume the change is for `ExecuteCommandAgentName` and instead pass the name of the managed agent this event is for as a parameter? just more future-proof

File: agent/engine/docker_task_engine.go
Comment (angelcar) [inline agent/engine/docker_task_engine.go:316]: Should we send the actual `error.Error()` here?

File: agent/engine/docker_task_engine.go
Comment (angelcar) [inline agent/engine/docker_task_engine.go:319]: Since this reason will be the customer facing reason, I wonder if "Task engine" is relevant. Would "Exec Command Agent Restarted" be more appropriate?

File: agent/engine/docker_task_engine.go
Comment (angelcar) [inline agent/engine/docker_task_engine.go:1317]: Wonder if we need to send a reason here.

File: agent/engine/task_manager.go
Comment (angelcar) [inline agent/engine/task_manager.go:533]: hmm, thinking about this more, this should be an error log, not debug :/

File: agent/engine/task_manager.go
Comment (angelcar) [inline agent/engine/task_manager.go:548]: Probably we don't need to modify these functions now that manged agent state changes are independent of containers

File: agent/eventhandler/task_handler.go
Comment (angelcar) [inline agent/eventhandler/task_handler.go:136]: debug level would be better? same for other logs below

File: agent/engine/task_manager.go
Comment (sharanyad) [inline agent/engine/task_manager.go:530]: I don't see the changes for emitting a stopped status event for managed agent? is that out of scope of this PR?

File: agent/eventhandler/task_handler.go
Comment (angelcar) [inline agent/eventhandler/task_handler.go:240]: this comment is sooo confusing, but I get it a similar one was already there

File: agent/eventhandler/task_handler_types.go
Comment (angelcar) [inline agent/eventhandler/task_handler_types.go:206]: are you addressing updating the sent status of managed agents as part of this PR?

File: agent/engine/task_manager.go
Comment (fierlion) [reply agent/engine/task_manager.go:548]: `func (mtask *managedTask) emitContainerEvent(task *apitask.Task, cont *apicontainer.Container, reason string)` in master/dev branches.  
https://github.com/aws/amazon-ecs-agent/blob/master/agent/engine/task_manager.go#L529

File: agent/eventhandler/task_handler_types.go
Comment (fierlion) [reply agent/eventhandler/task_handler_types.go:206]: no that will be a separate PR.  (This is to immediately unblock testing.)

File: agent/api/statechange.go
Comment (fierlion) [reply agent/api/statechange.go:156]: I'll add this to a later PR.

File: agent/ecs_client/model/api/api-2.json
Comment (fierlion) [reply agent/ecs_client/model/api/api-2.json:2118]: 1564     "ManagedAgentStateChanges":{

it is.

File: agent/eventhandler/task_handler.go
Comment (fierlion) [reply agent/eventhandler/task_handler.go:240]: agree.  I'll leave it for now.  We can refactor comments as we refactor code.

File: agent/engine/task_manager.go
Comment (fierlion) [reply agent/engine/task_manager.go]: It is out of scope of the current PR.

File: agent/engine/docker_task_engine.go
Comment (fierlion) [reply agent/engine/docker_task_engine.go]: This shows up in the reason field for describe-tasks -- would be nice to keep it terse.
aka
"Exec Command Agent Restarted"

File: agent/engine/task_manager.go
Comment (angelcar) [reply agent/engine/task_manager.go:548]: oh! nvm

File: agent/api/ecsclient/client.go
Comment (angelcar) [inline agent/api/ecsclient/client.go:403]: Let's do this a bit better and just do:
```golang
for i, managedAgentEvent := range change.ManagedAgents {
    if mgspl := client.buildManagedAgentStateChangePayload(managedAgentEvent); mgspl != nil {
        req.ManagedAgents = append(req.ManagedAgents, mgspl)
    }
}
```

File: agent/api/ecsclient/client.go
Comment (angelcar) [inline agent/api/ecsclient/client.go:428]: nit: this function looks prettier (more readable) like so:

```golang
func (client *APIECSClient) buildManagedAgentStateChangePayload(change api.ManagedAgentStateChange) *ecs.ManagedAgentStateChange {
	if !change.Status.ShouldReportToBackend() {
		seelog.Warnf("Not submitting unsupported managed agent state %s for container %s in task %s",
			status.String(), change.ContainerName, change.TaskArn)
		return nil
	}
	var trimmedReason *string
	if change.Reason != "" {
		trimmedReason := aws.String(trimString(change.Reason, ecsMaxReasonLength))
	}
	return &ecs.ManagedAgentStateChange{
		ManagedAgentName: aws.String(change.Name),
		ContainerName:    aws.String(change.ContainerName),
		Status:           aws.String(change.Status.String()),
		Reason:           trimmedReason,
	}
}
```

File: agent/eventhandler/task_handler_types.go
Comment (angelcar) [inline agent/eventhandler/task_handler_types.go:97]: do we need these comments?

**Review by angelcar — APPROVED** [review]

File: agent/eventhandler/task_handler_types.go
Comment (fierlion) [reply agent/eventhandler/task_handler_types.go:97]: We do. This section wouldn't make sense with what `make static-check` required. And it will change when I add the sent status.

File: agent/api/ecsclient/client.go
Comment (sharanyad) [inline agent/api/ecsclient/client.go:397]: please add include this in the unit tests to verify if managed agent changes are in the task state change.

File: agent/api/statechange.go
Comment (sharanyad) [inline agent/api/statechange.go:180]: missing unit tests

File: agent/ecs_client/model/ecs/api.go
Comment (sharanyad) [inline agent/ecs_client/model/ecs/api.go:9388]: is this check not required anymore?

File: agent/engine/task_manager.go
Comment (sharanyad) [inline agent/engine/task_manager.go:530]: missing unit tests for this method.

File: agent/eventhandler/task_handler.go
Comment (sharanyad) [inline agent/eventhandler/task_handler.go:139]: these could get noisy in the logs, should probably remove it. same for the one above switch statement and below ones

File: agent/eventhandler/task_handler_types.go
Comment (sharanyad) [inline agent/eventhandler/task_handler_types.go:100]: please remove comments before merging

File: agent/eventhandler/task_handler_types.go
Comment (sharanyad) [inline agent/eventhandler/task_handler_types.go:98]: could add a test case for this

File: agent/eventhandler/task_handler_types_test.go
Comment (sharanyad) [inline agent/eventhandler/task_handler_types_test.go:317]: please remove comments

File: agent/ecs_client/model/ecs/api.go
Comment (fierlion) [reply agent/ecs_client/model/ecs/api.go:9388]: this code is autogenerated from the agent/ecs_client/model/api/api-2.json change using `make gogenerate`

File: agent/eventhandler/task_handler.go
Comment (fierlion) [reply agent/eventhandler/task_handler.go:139]: removed all

File: agent/api/ecsclient/client.go
Comment (fierlion) [reply agent/api/ecsclient/client.go:428]: changed.

File: agent/api/statechange.go
Comment (fierlion) [reply agent/api/statechange.go:180]: agent/api/statechange_test.go has updated:
`func TestNewManagedAgentChangeEvent(t *testing.T) {`

Comment (fierlion) [reply agent/api/statechange.go:180]: added case for non-existent managed agent.

File: agent/engine/task_manager.go
Comment (fierlion) [reply agent/engine/task_manager.go:530]: this is being tested by the caller, `agent/engine/docker_task_engine_test.go` with checkManagedAgentEvents (in `/agent/engine/common_test.go`)
The test doesn't have any return value -- it's testable only by its side-effect: an event added to queue.

File: agent/api/statechange.go
Comment (sharanyad) [inline agent/api/statechange.go:184]: nit: we should change the log statement to print `managedAgentName` instead of hardcoding it as `execute command agent` (future proof)

**Review by sharanyad — APPROVED** [review]

File: agent/api/statechange.go
Comment (fierlion) [reply agent/api/statechange.go:184]: Agree -- I have a follow-up to fix the managed agent sent status.  will update there.

**Pull request merged**
